### PR TITLE
fix: Xplan validation erroneously throws warning on missing objects

### PR DESCRIPTION
**Pull request by emanuelaepure10**
Closes #4318

### CODE REVIEW

File: common/plugins/eu.esdihumboldt.hale.common.instance.graph/src/eu/esdihumboldt/hale/common/instance/graph/reference/ReferenceGraph.java [hunk 1]
```diff
@@ -235,24 +241,24 @@ public Boolean compute(LoopBundle<Vertex> loop) {
 					if (ref != null) {
 						result.add(ref);
 					}
-					else {
+					else if (!vtx.getId().equals(associated.getId())) {
```
Comment (stempler) [inline common/plugins/eu.esdihumboldt.hale.common.instance.graph/src/eu/esdihumboldt/hale/common/instance/graph/reference/ReferenceGraph.java:244]: So there were references from one instance to itself?

File: common/plugins/eu.esdihumboldt.hale.common.instance.graph/src/eu/esdihumboldt/hale/common/instance/graph/reference/ReferenceGraph.java [hunk 2]
```diff
@@ -235,24 +241,24 @@ public Boolean compute(LoopBundle<Vertex> loop) {
...
 						Iterable<Vertex> referers = associated.getVertices(Direction.IN);
 						Set<String> ids = new HashSet<>();
-						if (referers != null) {
-							for (Vertex referer : referers) {
-								Object ident = referer.getId();
-								if (ident != null) {
-									ids.add(ident.toString());
-								}
+
```
Comment (stempler) [inline common/plugins/eu.esdihumboldt.hale.common.instance.graph/src/eu/esdihumboldt/hale/common/instance/graph/reference/ReferenceGraph.java:247]: Any particular reason to remove the null check?

File: common/plugins/eu.esdihumboldt.hale.common.instance.graph/src/eu/esdihumboldt/hale/common/instance/graph/reference/ReferenceGraph.java [hunk 3]
```diff
@@ -235,24 +241,24 @@ public Boolean compute(LoopBundle<Vertex> loop) {
...
+						for (Vertex referer : referers) {
+							Object ident = referer.getId();
+							if (ident != null) {
+								ids.add(ident.toString());
 							}
 						}
+
 						if (ids.isEmpty()) {
-							log.warn("Encountered referenced object w/o associated instance: "
+							log.warn("Encountered referenced object without associated instance: "
```
Comment (stempler) [inline common/plugins/eu.esdihumboldt.hale.common.instance.graph/src/eu/esdihumboldt/hale/common/instance/graph/reference/ReferenceGraph.java:256]: Any particular reason to change the text?

File: common/plugins/eu.esdihumboldt.hale.common.instance.graph/src/eu/esdihumboldt/hale/common/instance/graph/reference/ReferenceGraph.java [hunk 4]
```diff
@@ -294,23 +300,36 @@ public void remove() {
...
 	private boolean partitioned = false;
 
+	private final T handleFirst;
+
 	/**
 	 * Create a new reference graph from the given instance collection.
 	 * 
 	 * @param inspector the instance inspector to use
 	 * @param instances the
+	 * @param handleFirst first to be handle
```
Comment (stempler) [inline common/plugins/eu.esdihumboldt.hale.common.instance.graph/src/eu/esdihumboldt/hale/common/instance/graph/reference/ReferenceGraph.java:310]: Maybe more detailed?
It should be clear that what is passed in here is an ID, and that it is optional and null may be passed in.

File: common/plugins/eu.esdihumboldt.hale.common.instance.graph/src/eu/esdihumboldt/hale/common/instance/graph/reference/ReferenceGraph.java (hunk 1 above)
Comment (emanuelaepure10) [reply common/plugins/eu.esdihumboldt.hale.common.instance.graph/src/eu/esdihumboldt/hale/common/instance/graph/reference/ReferenceGraph.java:244]: Yes and that was the case when actually the warning was given.

File: common/plugins/eu.esdihumboldt.hale.common.instance.graph/src/eu/esdihumboldt/hale/common/instance/graph/reference/ReferenceGraph.java (hunk 2 above)
Comment (emanuelaepure10) [reply common/plugins/eu.esdihumboldt.hale.common.instance.graph/src/eu/esdihumboldt/hale/common/instance/graph/reference/ReferenceGraph.java:247]: I haven't, it's just below.

Comment (stempler) [reply common/plugins/eu.esdihumboldt.hale.common.instance.graph/src/eu/esdihumboldt/hale/common/instance/graph/reference/ReferenceGraph.java:247]: No, `referers` is no longer checked for `null`.
Did you check the API and it is guaranteed to never be `null`?

File: common/plugins/eu.esdihumboldt.hale.common.instance.graph/src/eu/esdihumboldt/hale/common/instance/graph/reference/ReferenceGraph.java (hunk 1 above)
Comment (stempler) [reply common/plugins/eu.esdihumboldt.hale.common.instance.graph/src/eu/esdihumboldt/hale/common/instance/graph/reference/ReferenceGraph.java:244]: Did you identify how that can happen? I don't remember self-references from the test data that we looked at and even if there were I would expect the `ref != null` condition to be `true`.
Any idea? Just wondering if something is not correct when building the graph.

File: common/plugins/eu.esdihumboldt.hale.common.instance.graph/src/eu/esdihumboldt/hale/common/instance/graph/reference/ReferenceGraph.java (hunk 3 above)
Comment (stempler) [reply common/plugins/eu.esdihumboldt.hale.common.instance.graph/src/eu/esdihumboldt/hale/common/instance/graph/reference/ReferenceGraph.java]: Just curious

File: common/plugins/eu.esdihumboldt.hale.common.instance.graph/src/eu/esdihumboldt/hale/common/instance/graph/reference/ReferenceGraph.java (hunk 2 above)
Comment (emanuelaepure10) [reply common/plugins/eu.esdihumboldt.hale.common.instance.graph/src/eu/esdihumboldt/hale/common/instance/graph/reference/ReferenceGraph.java:247]: The `associated.getVertices(Direction.IN)` will return an empty iterable if there are no vertices.
I actually wanted to used this 
```
Set<String> ids = new HashSet<>();
for (Vertex referer : associated.getVertices(Direction.IN)) {
    Object ident = referer.getId();
    if (ident != null) {
        ids.add(ident.toString());
    }
}
```

File: common/plugins/eu.esdihumboldt.hale.common.instance.graph/src/eu/esdihumboldt/hale/common/instance/graph/reference/ReferenceGraph.java (hunk 1 above)
Comment (emanuelaepure10) [reply common/plugins/eu.esdihumboldt.hale.common.instance.graph/src/eu/esdihumboldt/hale/common/instance/graph/reference/ReferenceGraph.java:244]: Yes, the condition `vtx.getId().equals(associated.getId())` is `true` for the example attached to the issue many times, as well as for the sample example that we have created!

**Review by stempler — APPROVED** [review]

**Comment by stempler** [comment]
Related ticket is #890

**Comment by stempler** [comment]
@emanuelaepure10 I proposed a change for you to review in a fixup commit.

We had done a change together to try to have the first atomic part in the partitioning for each plan to include the plan. Only the change to first check the plan vertex in the reference graph was not enough, because that first atomic part was then skipped for later processing, since its size exceeded the maximum size desired for a part.
I modified the behavior to not skip the atomic part if the part that is currently prepared is empty. Like that we make sure that the first part retrieved actually contains the plan, and no other parts are being looked at (and thus we don't get any warnings).

**Review by emanuelaepure10 — COMMENTED** [review]
LGTM

**Pull request merged**